src/common/configIntelliSenseProvider.ts

```diff
@@ -0,0 +1,336 @@
+        // }
+
+        if (this.locationMatch(location, Constants.imgDeploymentManifestJsonPath)) {
+            const images: string[] = await this.getSlnImgPlaceholders(document.uri);
+            return this.getCompletionItems(images, document, position, location);
+        }
+
+        if (this.locationMatch(location, Constants.routeDeploymentManifestJsonPath)) {
+            const json = parser.parse(document.getText());
+            const modules: any = ((json.moduleContent.$edgeAgent || {})["properties.desired"] || {}).modules || {};
+            const moduleIds: string[] = Object.keys(modules);
+
+            const routeCompletionItem: vscode.CompletionItem = new vscode.CompletionItem(Constants.routeSnippetLabel);
+            routeCompletionItem.filterText = `\"${Constants.routeSnippetLabel}\"`;
+            routeCompletionItem.kind = vscode.CompletionItemKind.Snippet;
+            routeCompletionItem.detail = Constants.routeSnippetDetail;
+            routeCompletionItem.range = document.getWordRangeAtPosition(position);
+            routeCompletionItem.insertText = new vscode.SnippetString(this.getRouteSnippetString(moduleIds));
+            return [routeCompletionItem];
+        }
+    }
+
+    public async provideHover(document: vscode.TextDocument, position: vscode.Position): Promise<vscode.Hover> {
+        const location: parser.Location = parser.getLocation(document.getText(), document.offsetAt(position));
+
+        if (this.locationMatch(location, Constants.imgDeploymentManifestJsonPath)) {
+            const moduleToImageMap: Map<string, string> = new Map();
+            const imageToDockerfileMap: Map<string, string> = new Map();
+
+            try {
+                await Utility.setSlnModulesMap(path.dirname(document.uri.fsPath), moduleToImageMap, imageToDockerfileMap);
+
+                const node: parser.Node = location.previousNode;
+                const imageUrl: string = node.value;
+                const imageUrlUnwrapped: string = imageUrl.slice(2, -1); // remove the wrapping "${" and "}"
+                const image = moduleToImageMap.get(imageUrlUnwrapped);
+                if (image) {
+                    const dockerfile: string = imageToDockerfileMap.get(image);
+                    const dockerfileContent: string = await fse.readFile(dockerfile, "utf-8");
+                    const range: vscode.Range = this.getNodeRange(document, node);
+                    return new vscode.Hover({ language: "dockerfile", value: dockerfileContent }, range);
+                }
+            } catch {
+                return undefined;
+            }
+        }
+
+        return undefined;
+    }
+
+    public async provideDefinition(document: vscode.TextDocument, position: vscode.Position): Promise<vscode.Definition> {
+        const location: parser.Location = parser.getLocation(document.getText(), document.offsetAt(position));
```

Comment:
duplicate code between provideDefinition and provideHover, extract an utility instead?

Reply:
Fixed. Thanks.